src/game/common/statemachine.cpp

```diff
@@ -0,0 +1,716 @@
+#ifdef GAME_DEBUG_STRUCTS
+                    if (Get_Machine()->Get_Wants_Debug_Output()) {
+                        const char *description;
+
+                        if (info.description != nullptr) {
+                            description = info.description;
+                        } else {
+                            description = "[no description]";
+                        }
+
+                        captainslog_debug("%d '%s' -- '%s' condition '%s' returned true!",
+                            g_theGameLogic->Get_Frame(),
+                            Get_Machine_Owner()->Get_Template()->Get_Name().Str(),
+                            Get_Machine()->Get_Name().Str(),
+                            description);
+                    }
+#endif
+
+                    if (info.to_state_id == EXIT_MACHINE_WITH_SUCCESS) {
+                        return STATE_SUCCESS;
+                    } else if (info.to_state_id == EXIT_MACHINE_WITH_FAILURE) {
+                        return STATE_FAILURE;
+                    } else {
+                        return Get_Machine()->Internal_Set_State(info.to_state_id);
+                    }
+                }
+            }
+        }
+
+        return status;
+    }
+}
+
+StateReturnType State::Friend_Check_For_Transitions(StateReturnType status)
+{
+    static int checkfortransitionsnum;
+    StIncrementer incrementer(&checkfortransitionsnum);
+
+    if (checkfortransitionsnum >= 20) {
+        // assert disabled, the game ends up with checkfortransitionsnum >= 20
+#if 0
+        captainslog_dbgassert(checkfortransitionsnum < 20, "checkfortransitionsnum is > 20");
+#endif
+        return STATE_FAILURE;
+    }
+
+    captainslog_dbgassert(status <= STATE_CONTINUE, "Please handle sleep states prior to this");
+
+    switch (status) {
+        case STATE_FAILURE: {
+            if (m_failureStateID == EXIT_MACHINE_WITH_SUCCESS) {
+                Get_Machine()->Internal_Set_State(MACHINE_DONE_STATE_ID);
+                return STATE_SUCCESS;
+            } else if (m_failureStateID == EXIT_MACHINE_WITH_FAILURE) {
+                Get_Machine()->Internal_Set_State(MACHINE_DONE_STATE_ID);
+                return STATE_FAILURE;
+            } else {
+                return Get_Machine()->Internal_Set_State(m_failureStateID);
+            }
+        } break;
+
+        case STATE_SUCCESS: {
+            if (m_successStateID == EXIT_MACHINE_WITH_SUCCESS) {
+                Get_Machine()->Internal_Set_State(MACHINE_DONE_STATE_ID);
+                return STATE_SUCCESS;
+            } else if (m_successStateID == EXIT_MACHINE_WITH_FAILURE) {
+                Get_Machine()->Internal_Set_State(MACHINE_DONE_STATE_ID);
+                return STATE_FAILURE;
+            } else {
+                return Get_Machine()->Internal_Set_State(m_successStateID);
+            }
+        } break;
+
+        case STATE_CONTINUE: {
+            if (!!m_transitions.empty()) {

```

Comment:
Condition is double inverse, so its wrong, but also obsolete. Can be removed.

Reply:
fixed.